binlog_writer.go

```diff
@@ -49,14 +57,23 @@ func (b *BinlogWriter) Run() {
 					// Channel is closed, finish writing batch.
 					wantMoreEvents = false
 				}
+			case <-ctx.Done():
+				b.logger.Info("cancellation received, terminating goroutine abruptly")
+				return
 			default: // Nothing in the buffer so just write it
 				wantMoreEvents = false
 			}
 		}
 
-		err := WithRetries(b.WriteRetries, 0, b.logger, "write events to target", func() error {
-			return b.writeEvents(batch)
+		err := WithRetriesContext(ctx, b.WriteRetries, 0, b.logger, "write events to target", func() error {
+			return b.writeEvents(ctx, batch)
 		})
+
+		if err == context.Canceled {
+			b.logger.Info("cancellation received, terminating goroutine abruptly")
+			return
+		}
```

Comment:
Hmmmm.. this seems very sketchy. I don't believe the binlog writer should take a context. But rather it should shut down naturally due to its buffer channel being closed by the Ferry/BinlogStreamer.  Otherwise, it will lose writes that we assumed would complete 

https://github.com/Shopify/ghostferry/blob/5e461ead1b6104d83fc736ece2b426e354a7cbe4/ferry.go#L279

https://github.com/Shopify/ghostferry/blob/5e461ead1b6104d83fc736ece2b426e354a7cbe4/ferry.go#L341-L343

Reply:
Good catch here. I overlooked the fact that we're buffering the binlog events to write here.

Reply:
Actually this is an issue with the panic based interruption as well. We have no idea as is, if a binlog has been written to the database or not in the binlog streamer.